File: zc.py
#!/usr/bin/env python3

from dataclasses import dataclass
from enum import IntEnum
import operator as op

class Array(list): pass
class Identifier(str): pass
class Keyword(str): pass
class Integer(int): pass
class Float(float): pass
class String(str): pass
Code_Data = Identifier | Keyword | Integer | Float | String | Array
@dataclass
class Code:
  loc: int
  data: Code_Data

def parse_code(s, p):
  level_stack = []
  stack = []
  while True:
    start = p
    newline_was_skipped = False
    while True:
      while p < len(s) and s[p].isspace():
        if s[p] == '\n': newline_was_skipped = True
        p += 1
      if p < len(s) and s[p] == ';':
        while p < len(s) and s[p] != '\n': p += 1
        continue
      break
    first_exp_of_line = newline_was_skipped or start == 0
    if p >= len(s): break

    if first_exp_of_line and s[p] != '(':
      level_stack.append(True)
      stack.append(Code(p, Array()))

    start = p
    if s[p] == '(':
      p += 1
      level_stack.append(False)
      stack.append(Code(p, Array()))
    elif s[p] == ')':
      p += 1

      while len(level_stack) > 0 and level_stack[-1]: level_stack.pop()
      if len(level_stack) <= 0: print("Missing (."); exit(1)

      level_stack.pop()
      popped = stack.pop()
      (stack[-1].data if len(stack) > 0 else stack).append(popped)
    elif s[p].isdigit():
      while p < len(s) and s[p].isdigit(): p += 1
      (stack[-1].data if len(stack) > 0 else stack).append(Code(start, Integer(s[start:p])))
    elif s[p] == "'":
      p += 1
      code, next_pos = parse_code(s, p)
      assert code is not None
      p = next_pos
      (stack[-1].data if len(stack) > 0 else stack).append(Code(start, Array([Code(start, Identifier("$codeof")), code])))
    else:
      while p < len(s) and (not s[p].isspace() and s[p] != ')'): p += 1
      (stack[-1].data if len(stack) > 0 else stack).append(Code(start, Identifier(s[start:p])))

    peek = p
    while peek < len(s) and s[peek] != '\n' and s[peek] != ';' and s[peek].isspace(): peek += 1
    if peek >= len(s) or s[peek] == '\n' or s[peek] == ';':
      while len(level_stack) > 0 and level_stack[-1]: level_stack.pop()

    if len(level_stack) == 0: break
  if len(level_stack) != 0: print("Missing )."); exit(1)
  assert len(stack) <= 1
  return stack.pop() if len(stack) > 0 else None, p

def print_code(code, show_kinds=False):
  if show_kinds: print(code.__class__.__name__ + "[", end="")
  if isinstance(code, Identifier | Integer): print(code, end="")
  elif isinstance(code, Array):
    print("(", end="")
    for c in code: print_code(c, show_kinds)
    print(")", end="")
  else: raise NotImplementedError(type(code))
  if show_kinds: print("]", end="")

class Type: pass
class Type_Procedure_Calling_Convention(IntEnum):
  DEFAULT = 0
  Z = 1
  C = 2
@dataclass
class Type_Procedure(Type):
  parameter_types: list[Type]
  parameter_names: list[str]
  return_type: Type
  callconv: Type_Procedure_Calling_Convention

type_type = Type()
type_code = Type()
type_void = Type()
type_bool = Type()
type_comptime_int = Type()

def print_type(type_):
  if type_ == type_type: print("($type 'TYPE)", end="")
  elif type_ == type_code: print("($type 'CODE)", end="")
  elif type_ == type_void: print("($type 'VOID)", end="")
  elif type_ == type_bool: print("($type 'BOOL)", end="")
  elif type_ == type_comptime_int: print("($type 'COMPTIME_INTEGER)", end="")
  else: raise NotImplementedError(type(type_))

@dataclass
class Value:
  type: Type
  data: Code_Data

value_true = Value(type_bool, None)
value_false = Value(type_bool, None)
value_void = Value(type_void, None)

@dataclass
class Env_Entry:
  constant: bool
  public: bool
  value: Value

class Env:
  def __init__(self):
    self.parent = None
    self.data = {}
  def find(self, key):
    if key in self.data: return self.data[key]
    if self.parent is not None: return self.parent.find(key)
    return None

def is_truthy(value):
  if value.type == type_bool: return value != value_false
  elif value.type == type_comptime_int: return value.data != 0 # TODO: decide if this should be a thing. (probably not)
  else: raise NotImplementedError(value.type)

def is_equal(a, b):
  if a.type == type_comptime_int and b.type == type_comptime_int: return a.data == b.data
  raise NotImplementedError(type(a), type(b))

def eval_code(code, env):
  assert isinstance(code, Code)
  if not isinstance(code.data, Array):
    if isinstance(code.data, Identifier):
      entry = env.find(code.data)
      if entry is None: print(f"Identifier '{code.data}' not found in environment."); exit(1)
      return entry.value
    elif isinstance(code.data, Integer):
      value = Value(type_comptime_int, code.data)
      return value
    else: raise NotImplementedError(type(code.data))
  op_code, *arg_codes = code.data
  if op_code.data == Identifier("$define"):
    name_code, value_code = arg_codes
    name = eval_code(name_code, env)
    assert name.type == type_code and isinstance(name.data, Identifier)
    assert name.data not in env.data
    env.data[name.data] = Env_Entry(constant=True, public=False, value=eval_code(value_code, env))
    return value_void
  elif op_code.data == Identifier("$codeof"):
    if len(arg_codes) != 1: print("$codeof expects one argument."); exit(1)
    return Value(type_code, arg_codes[0].data)
  elif op_code.data == Identifier("$insert"):
    if len(arg_codes) != 1: print("$insert expects one argument (for now)."); exit(1)
    value = eval_code(arg_codes[0], env)
    if value.type != type_code: print("$insert expects argument one to be of type ($type 'CODE)."); exit(1)
    return eval_code(Code(op_code.loc, value.data), env)
  elif op_code.data == Identifier("$if"):
    test, conseq, *alt = arg_codes
    assert len(alt) <= 1
    return eval_code(conseq, env) if is_truthy(eval_code(test, env)) else (eval_code(alt[0], env) if len(alt) != 0 else value_void)
  elif op_code.data == Identifier("$operator"):
    operator_code, *rest_codes = arg_codes
    operator = eval_code(operator_code, env)
    assert operator.type == type_code and isinstance(operator.data, Identifier)
    if operator.data == Identifier("=="):
      result = True
      arg1 = eval_code(rest_codes[0], env) if len(rest_codes) != 0 else None
      for arg_code in rest_codes[1:]:
        arg2 = eval_code(arg_code, env)
        result = result and is_equal(arg1, arg2)
      return value_true if result else value_false
    elif operator.data == Identifier("*"):
      result = 1
      if len(rest_codes) > 0:
        arg1 = eval_code(rest_codes[0], env)
        assert arg1.type == type_comptime_int
        result = arg1.data
      for arg_code in rest_codes[1:]:
        arg2 = eval_code(arg_code, env)
        assert arg2.type == type_comptime_int
        result *= arg2.data
      return Value(type_comptime_int, Integer(result))
    raise NotImplementedError()
  else:
    proc = eval_code(op_code, env)
    if not isinstance(proc.type, Type_Procedure):
      print(f"Attempted to call non-procedure of type '", end="")
      print_type(proc.type)
      print("'.")
      exit(1)
    pargs = [eval_code(arg_code, env) for arg_code in arg_codes]
    return proc(*pargs)

def print_value(value):
  if isinstance(value.data, Code_Data): print_code(value.data)
  elif value == value_true: print("($cast ($type 'BOOL) 1)", end="")
  elif value == value_false: print("($cast ($type 'BOOL) 0)", end="")
  else: raise NotImplementedError(type(value.data))

import sys
if len(sys.argv) == 1:
  env = Env()
  while True:
    i = input("> ")
    pos = 0
    if i in ["exit", "quit"] : break
    while True:
      code, next_pos = parse_code(i, pos)
      if code is None: break
      pos = next_pos
      result = eval_code(code, env)
      print("=> ", end="")
      print_value(result)
      print("")
else:
  with open(sys.argv[1]) as f: src = f.read()
  env = Env()
  pos = 0
  while True:
    code, next_pos = parse_code(src, pos)
    if code is None: break
    pos = next_pos
    eval_code(code, env)

  print("=====ENVIRONMENT=====")
  for key in env.data:
    print(key + ': ', end="")
    print_value(env.data[key].value)
    print("")

File: zc.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>
#include <string.h>
#include <ctype.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;
typedef size_t usize;

typedef struct {
  usize count;
  u8* data;
} slice_u8;

typedef slice_u8 string;

slice_u8 permanent_storage;
usize permanent_storage_used;

#define New(T) ((T*) push_(sizeof(T)))
#define Make(N, T) ((T*) push_(sizeof(T) * (N)))
void* push_(size_t size) {
  u8* result = permanent_storage.data + permanent_storage_used;
  permanent_storage_used += size + 15;
  permanent_storage_used &= ~0xF;
  return result;
}

void die(char* msg, ...) {
  va_list ap;
  va_start(ap, msg);
  vfprintf(stderr, msg, ap);
  va_end(ap);
  exit(EXIT_FAILURE);
}

typedef enum Exp_Kind {
  EXP_IDENTIFIER,
  EXP_KEYWORD,
  EXP_INTEGER,
  EXP_FLOAT,
  EXP_STRING,
  EXP_ARRAY,
} Exp_Kind;

typedef struct slice_ExpPtr {
  usize count;
  struct Exp** data;
} slice_ExpPtr;

typedef struct Exp {
  usize loc;
  Exp_Kind kind;
  union {
    string as_atom;
    slice_ExpPtr as_array;
  } u;
} Exp;

#define stack_push(X, Y) { (X)->data[(X)->count++] = (Y); }
#define stack_pop(X) ((X)->data[--(X)->count])
Exp* parse_exp(string s, u32 p, u32* next_p, bool first_exp_in_line) {
  u32 level = 0;
  static slice_ExpPtr stack;
  stack.count = 0;
  if (stack.data == NULL) stack.data = Make(64, Exp*);
  bool implicit_parentheses = false;
  bool is_quoting = false;
  for (;;) {
    bool newline_was_skipped = false;
    for (;;) {
      while (p < s.count && isspace(s.data[p])) {
        if (s.data[p] == '\n') newline_was_skipped = true;
        p += 1;
      }
      if (p < s.count && s.data[p] == ';') {
        while (p < s.count && s.data[p] != '\n') p += 1;
        continue;
      }
      break;
    }
    u32 start = p;
    if (implicit_parentheses && level == 1 && newline_was_skipped) {
      level -= 1;
    } else if (p >= s.count) {
      break;
    } else if ((first_exp_in_line && level == 0) || s.data[p] == '(') {
      if (s.data[p] == '(') p += 1;
      else implicit_parentheses = true;
      level += 1;
      Exp* array = New(Exp);
      array->loc = start;
      array->kind = EXP_ARRAY;
      array->u.as_array.data = Make(64, Exp*); // TODO(dfra): wow be smarter than this please
      stack_push(&stack, array);
    } else if (s.data[p] == ')') {
      p += 1;
      if (level < 1) die("%d: Unmatched )\n", p);
      level -= 1;
      assert(stack.count > 0);
      Exp* popped = stack_pop(&stack);
      stack_push(stack.count > 0 ? &stack.data[stack.count - 1]->u.as_array : &stack, popped);
    } else if (s.data[p] == '\'') {
      is_quoting = true;
      p += 1;
      level += 1;
      Exp* array = New(Exp);
      array->loc = start;
      array->kind = EXP_ARRAY;
      array->u.as_array.count = 0;
      array->u.as_array.data = Make(64, Exp*);
      Exp* identifier = New(Exp);
      identifier->loc = start;
      identifier->kind = EXP_IDENTIFIER;
      identifier->u.as_atom.count = 7;
      identifier->u.as_atom.data = "$codeof";
      stack_push(&array->u.as_array, identifier);
      stack_push(&stack, array);
      continue;
    } else if (isdigit(s.data[p])) {
      while (p < s.count && isdigit(s.data[p])) p += 1;
      Exp* identifier = New(Exp);
      identifier->loc = start;
      identifier->kind = EXP_INTEGER;
      identifier->u.as_atom.data = s.data + start;
      identifier->u.as_atom.count = p - start;
      stack_push(stack.count > 0 ? &stack.data[stack.count - 1]->u.as_array : &stack, identifier);
    } else {
      while (p < s.count && (!isspace(s.data[p]) && s.data[p] != ')')) p += 1;
      Exp* identifier = New(Exp);
      identifier->loc = start;
      identifier->kind = EXP_IDENTIFIER;
      identifier->u.as_atom.data = s.data + start;
      identifier->u.as_atom.count = p - start;
      stack_push(stack.count > 0 ? &stack.data[stack.count - 1]->u.as_array : &stack, identifier);
    }
    first_exp_in_line = false;
    if (is_quoting) {
      level -= 1;
      Exp* popped = stack_pop(&stack);
      stack_push(stack.count > 0 ? &stack.data[stack.count - 1]->u.as_array : &stack, popped);
      is_quoting = false;
    }
    if (level == 0) break;
  }
  if (level != 0) die("%d: Unmatched (\n", p);
  assert(stack.count <= 1);
  *next_p = p;
  return stack.count > 0 ? stack.data[stack.count - 1] : NULL;
}

void print_exp(Exp* x, string s, u32 level, bool show_kinds) {
  if (show_kinds) printf("%d[", x->kind);
  switch (x->kind) {
    case EXP_IDENTIFIER:
    case EXP_KEYWORD:
    case EXP_INTEGER:
    case EXP_FLOAT:
    case EXP_STRING:
      printf("%.*s", (int) x->u.as_atom.count, x->u.as_atom.data);
      break;
    case EXP_ARRAY:
      printf("(");
      for (usize i = 0; i < x->u.as_array.count; i += 1) {
        Exp* child = x->u.as_array.data[i];
        assert(child != NULL);
        print_exp(child, s, level + 1, show_kinds);
        if (i != x->u.as_array.count - 1) printf(" ");
      }
      printf(")");
      break;
  }
  if (show_kinds) printf("]");
  if (level == 0) printf("\n");
}

typedef struct Type_Integer {
  u8 bits;
  bool is_signed;
} Type_Integer;

typedef enum Type_Pointer_Kind {
  POINTER_SINGLE,
  POINTER_MANY,
  POINTER_SLICE,
} Type_Pointer_Kind;

typedef struct Type_Pointer {
  Type_Pointer_Kind kind;
  struct Type* child;
} Type_Pointer;

typedef enum Type_Kind {
  TYPE_TYPE,
  TYPE_CODE,
  TYPE_VOID,
  TYPE_BOOL,
  TYPE_COMPTIME_INT,
  TYPE_COMPTIME_FLOAT,
  TYPE_INTEGER,
  TYPE_POINTER,
  TYPE_PROCEDURE,
} Type_Kind;

typedef struct Type {
  Type_Kind kind;
  union {
    Type_Integer as_integer;
    Type_Pointer as_pointer;
  } u;
} Type;

Type* type_code;
Type* type_void;
Type* type_comptime_int;
Type* type_comptime_float;
Type* type_u8;
Type* type_slice_u8;

typedef enum Value_Kind {
  VALUE_EXP,
  VALUE_TYPE,
  VALUE_PROCEDURE,
  VALUE_VOID,
} Value_Kind;

typedef struct Value {
  Type* type;
  Value_Kind kind;
  union {
    Exp* as_exp;
    Type* as_type;
  } u;
} Value;

Value* value_void;

typedef struct Env_Entry {
  bool constant;
  bool public;
  string key;
  Value* value;
} Env_Entry;

typedef struct slice_Env_EntryPtr {
  usize count;
  Env_Entry** data;
} slice_Env_EntryPtr;

typedef struct Env {
  struct Env* parent;
  slice_Env_EntryPtr table;
} Env;

#define string_equal_literal(A, B) string_equal((A), ((string){.count = sizeof(B) - 1, .data = (B)}))
bool string_equal(string a, string b) {
  if (a.count != b.count) return false;
  if (strncmp(a.data, b.data, a.count) != 0) return false;
  return true;
}

Env_Entry* table_find_pointer(slice_Env_EntryPtr* table, string key) {
  if (key.count == 0) return NULL;
  for (Env_Entry** entry = table->data; entry < table->data + table->count; entry += 1) {
    if (string_equal(key, (*entry)->key)) {
      return *entry;
    }
  }
  return NULL;
}

Env_Entry* env_find(Env* env, string identifier) {
  Env_Entry* entry = table_find_pointer(&env->table, identifier);
  if (entry != NULL) return entry;
  if (env->parent != NULL) return env_find(env->parent, identifier);
  return NULL;
}

void env_add(Env* env, string key, Value* value, bool constant, bool public) {
  Env_Entry* entry = New(Env_Entry);
  entry->constant = constant;
  entry->public = public;
  entry->key = key;
  entry->value = value;
  env->table.data[env->table.count++] = entry;
}

Value* eval_exp(Exp* x, Env* env, string s) {
  switch (x->kind) {
    case EXP_IDENTIFIER: {
      Env_Entry* entry = env_find(env, x->u.as_atom);
      if (entry == NULL) die("%zu: I failed to find %.*s in the current environment.\n", x->loc, (int) x->u.as_atom.count, x->u.as_atom.data);
      return entry->value;
    }
    case EXP_KEYWORD: {
      Value* value = New(Value);
      value->type = type_code;
      value->kind = VALUE_EXP;
      value->u.as_exp = x;
      return value;
    }
    case EXP_INTEGER: {
      Value* value = New(Value);
      value->type = type_comptime_int;
      value->kind = VALUE_EXP;
      value->u.as_exp = x;
      return value;
    }
    case EXP_FLOAT: {
      Value* value = New(Value);
      value->type = type_comptime_float;
      value->kind = VALUE_EXP;
      value->u.as_exp = x;
      return value;
    }
    case EXP_STRING: {
      Value* value = New(Value);
      value->type = type_slice_u8;
      value->kind = VALUE_EXP;
      value->u.as_exp = x;
      return value;
    }
    case EXP_ARRAY: {
      slice_ExpPtr* array = &x->u.as_array;
      if (array->count == 0) die("%zu: You tried to a call a procedure but didn't specify its name.\n", x->loc);

      Exp* op_exp = array->data[0];
      if (string_equal_literal(op_exp->u.as_atom, "$codeof")) {
        if (array->count != 2) die("%zu: $codeof expects exactly one argument.\n", x->loc);
        Value* value = New(Value);
        value->type = type_code;
        value->kind = VALUE_EXP;
        value->u.as_exp = array->data[1];
        return value;
      } else if (string_equal_literal(op_exp->u.as_atom, "$insert")) {
        if (array->count != 2) die("%zu: $insert expects exactly one argument (for now).\n", x->loc);
        return eval_exp(array->data[1], env, s);
      } else if (string_equal_literal(op_exp->u.as_atom, "$define")) {
        if (array->count != 3) die("%zu: $define expects exactly two arguments (for now).\n", x->loc);
        Exp* key_exp = array->data[1];
        Value* key = eval_exp(key_exp, env, s);
        if (key->kind != VALUE_EXP || key->u.as_exp->kind != EXP_IDENTIFIER) die("%zu: $define expects an identifier as its first argument.\n", x->loc);
        env_add(env, key->u.as_exp->u.as_atom, eval_exp(array->data[2], env, s), true, false);
        return value_void;
      } else {
        Value* maybe_procedure = eval_exp(op_exp, env, s);
        if (maybe_procedure->type->kind != TYPE_PROCEDURE) die("%zu: You tried to call something that wasn't a procedure.\n", op_exp->loc);
        assert(false); // unimplmented
        return NULL;
      }
    }
  }
}

void print_type(Type* type) {
  switch (type->kind) {
    case TYPE_TYPE: printf("($type 'TYPE)"); break;
    case TYPE_CODE: printf("($type 'CODE)"); break;
    case TYPE_VOID: printf("($type 'VOID)"); break;
    case TYPE_BOOL: printf("($type 'BOOL)"); break;
    case TYPE_COMPTIME_INT: printf("($type 'COMPTIME_INT)"); break;
    case TYPE_COMPTIME_FLOAT: printf("($type 'COMPTIME_FLOAT)"); break;
    case TYPE_INTEGER: printf("($type 'INTEGER '(#bits %d #signed ($cast ($type 'BOOL) %d)))", type->u.as_integer.bits, type->u.as_integer.is_signed); break;
    case TYPE_POINTER: printf("($type 'POINTER '(#kind %s #child ", type->u.as_pointer.kind == POINTER_SINGLE ? "'SINGLE" : type->u.as_pointer.kind == POINTER_MANY ? "'MANY" : "'SLICE"); print_type(type->u.as_pointer.child); printf("))"); break;
    case TYPE_PROCEDURE: printf("not implemented"); break;
  }
}

void print_value(Value* value, string s) {
  switch (value->kind) {
    case VALUE_VOID: printf("($cast ($type .VOID) 0)\n"); break;
    case VALUE_TYPE: print_type(value->u.as_type); printf("\n"); break;
    case VALUE_EXP: print_exp(value->u.as_exp, s, 0, false); break;
    case VALUE_PROCEDURE: printf("proc unimplemented\n"); break;
  }
}

string read_entire_file(char* filepath) {
  string result = {0};
  FILE* f = fopen(filepath, "rb");
  if (f) {
    fseek(f, 0, SEEK_END);
    long len = ftell(f);
    rewind(f);
    u8* buf = permanent_storage.data + permanent_storage_used;
    fread(buf, 1, len, f);
    buf[len] = 0;
    permanent_storage_used += len + 1;
    fclose(f);
    result.data = buf;
    result.count = len;
  }
  return result;
}

int main(int argc, char** argv) {
  permanent_storage.count = 1 * 1024 * 1024 * 1024;
  permanent_storage.data = calloc(permanent_storage.count, 1);
  assert(permanent_storage.data != NULL);

  if (argc == 0) die("I received no arguments from the command line.\n");
  if (argc == 1) die("Please provide me a file to compile, like this: `%s your_main_file.z`\n", argv[0]);

  type_code = &(Type){.kind = TYPE_CODE};
  type_void = &(Type){.kind = TYPE_VOID};
  type_comptime_int = &(Type){.kind = TYPE_COMPTIME_INT};
  type_comptime_float = &(Type){.kind = TYPE_COMPTIME_FLOAT};
  type_u8 = &(Type){.kind = TYPE_INTEGER, .u.as_integer = {.bits = 8, .is_signed = false}};
  type_slice_u8 = &(Type){.kind = TYPE_POINTER, .u.as_pointer = {.kind = POINTER_SLICE, .child = type_u8}};

  value_void = &(Value){.kind = VALUE_VOID, .type = type_void};

  string src = read_entire_file(argv[1]);
  // printf("%.*s", (int) src.count, src.data);
  u32 pos = 0;
  Env env = {0};
  env.table.data = Make(512, Env_Entry*);
  for (;;) {
    u32 next_pos;
    Exp* exp = parse_exp(src, pos, &next_pos, true);
    if (exp == NULL) break;
    pos = next_pos;
    print_exp(exp, src, 0, false);
    Value* value = eval_exp(exp, &env, src);
    if (value != value_void) print_value(value, src);
  }
  for (Env_Entry** entry = env.table.data; entry < env.table.data + env.table.count; entry += 1) {
    printf("%.*s: ", (int) (*entry)->key.count, (*entry)->key.data);
    print_value((*entry)->value, src);
  }

  return EXIT_SUCCESS;
}
